### PR TITLE
Fixed Improper Method Call: Replaced `NotImplementedError`

### PR DESCRIPTION
**Pull request by fazledyn-or**
## Details
While triaging your project, our bug fixing tool generated the following message(s)-

> In file: [spectral_cube.py](https://github.com/radio-astro-tools/spectral-cube/blob/master/spectral_cube/spectral_cube.py#L2299), class: BaseSpectralCube, there is a special method `__floordiv__` that raises a NotImplementedError. If a special method supporting a binary operation is not implemented it should return NotImplemented. On the other hand, NotImplementedError should be raised from abstract methods inside user defined base classes to indicate that derived classes should override those methods. iCR suggested that the special method `__floordiv__` should return NotImplemented instead of raising an exception. An example of how NotImplemented helps the interpreter support a binary operation is here.

### Related Documentation
- [docs.python.org - NotImplemented](https://docs.python.org/3/library/constants.html#NotImplemented)
- [docs.python.org - NotImplementedError](https://docs.python.org/3/library/exceptions.html#NotImplementedError)
- [docs.python.org - Implementing Arithmetic Operators](https://docs.python.org/3/library/numbers.html#implementing-the-arithmetic-operations)

## Changes
- Replaced `NotImplementedError` with `NotImplemented`

## Previously Found & Fixed
- https://www.github.com/SciTools/iris/pull/5544
- https://www.github.com/cupy/cupy/pull/7900
- https://www.github.com/ethereum/web3.py/pull/3080

## CLA Requirements
*This section is only relevant if your project requires contributors to sign a Contributor License Agreement (CLA) for external contributions.*

All contributed commits are already automatically signed off.

> The meaning of a signoff depends on the project, but it typically certifies that committer has the rights to submit this work under the same license and agrees to a Developer Certificate of Origin (see [https://developercertificate.org/](https://developercertificate.org/) for more information).
\- [Git Commit SignOff documentation](https://developercertificate.org/)

## Sponsorship and Support
This work is done by the security researchers from OpenRefactory and is supported by the [Open Source Security Foundation (OpenSSF)](https://openssf.org/): [Project Alpha-Omega](https://alpha-omega.dev/). Alpha-Omega is a project partnering with open source software project maintainers to systematically find new, as-yet-undiscovered vulnerabilities in open source code - and get them fixed – to improve global software supply chain security.

The bug is found by running the Intelligent Code Repair (iCR) tool by OpenRefactory and then manually triaging the results.

### CODE REVIEW

**Comment by keflavich** [comment]
I'm a bit confused by this recommendation.  Why isn't `NotImplentedError` acceptable?  I don't see an explanation.

Also, though, in this case, the recommendation they give is:
```
Note It should not be used to indicate that an operator or method is not meant to be supported at all – in that case either leave the operator / method undefined or, if a subclass, set it to [None](https://docs.python.org/3/library/constants.html#None).
```
i.e., `__floordiv__` should be set to `None`, not `return NotImplemented`.  The latter is if there is a type-specific `NotImplemented`.

**Comment by fazledyn-or** [comment]
The reason behind using `NotImplemented` instead of `NotImplementedError` is according to the documentation.

Let's say we have `__eq__` method in a `MyClass` class that raises `NotImplementedError`. In that case, the following code would result in an exception - 
```python
a = MyClass()
b = a
print(a == b)
```
Whereas, the correct output should've been `True`. If we change `NotImplementedError` to `NotImplemented`, the Python interpreter does the comparison for us.

Also, using `NotImplemented` results in TypeError if the operation isn't supported at all. Whereas using `None` results in `None`, with no extra messages that helps the developer.

Running the following program gives us the output - 
```python
class Foo:
	def __floordiv__(self, other):
		raise NotImplementedError("Floor-division is not supported")

class Bar:
	def __floordiv__(self, other):
		None     # or pass

class Tone:
	def __floordiv__(self, other):
		return NotImplemented

try:
	a = Foo()
	b = Foo()
	print(a // b)
except Exception as e:
	print(e)

try:
	x = Bar()
	y = Bar()
	print(x // y)
except Exception as e:
	print(e)

try:
	m = Tone()
	n = Tone()
	print(m // n)
except Exception as e:
	print(e)
```
```
Floor-division is not supported
None
unsupported operand type(s) for //: 'Tone' and 'Tone'
```

Which is why, I think using `NotImplemented` rather than `NotImplementedError` or `None` is a better option.

**Comment by fazledyn-or** [comment]
> I'm still not convinced I see the case for `NotImplemented` over raising an exception; even in the case you gave above, it's not clear that `a==b` _should_ be True (I think it is True because you're actually doing `NotImplemented == NotImplemented`?).

Umm, I don't think so.

Putting `return NotImplemented` for `__eq__` method for `MyClass` hands over the comparison to the Python interpreter. Then the interpreter compares it (a and b are the same variable, just different identifiers). For example, we can run the following code -
```python
class Foo:
        def __eq__(self, other):
                return NotImplemented

class Bar:
        def __eq__(self, other):
                return NotImplemented

a = Foo()
b = Bar()
print(a == b)
```
```
False
```

Even though both class are returning `NotImplemented`, they aren't equal at all. Because the underlying Python interpreter knows that they're of different types. Which isn't actually what you thought `NotImplemented` does.

**Comment by keflavich** [comment]
OK, that's helpful.

I suppose the best default is for an object to equal itself.  There are exceptional cases where you might not want that (e.g., `nan`), but the example above isn't one of them.

**Comment by keflavich** [comment]
The tests need to be updated to accommodate the change in behavior, since we were testing for `NotImplementedError` before.

**Comment by fazledyn-or** [comment]
Okay, I'll update the tests ASAP and push.

**Comment by fazledyn-or** [comment]
I've updated the tests. However, the test `test_floordiv_fails` kept failing, as a result- I had to remove the pytest context for it. Please have a look and let me know if it's correct or not.

File: spectral_cube/tests/test_spectral_cube.py [hunk 1]
```diff
@@ -545,24 +545,17 @@ def test_div_cubes(self):
...
             c2 = self.c1 // value
         self.c1 = self.d1 = None
 
     @pytest.mark.parametrize(('value'),(1,1.0,2,2.0)*u.K)
     def test_floordiv_fails(self, value):
-        with pytest.raises(NotImplementedError,
-                           match=re.escape("Floor-division (division with truncation) "
-                                           "is not supported.")):
-            c2 = self.c1 // value
+        c2 = self.c1 // value
```
[the remark this answers is not in the document]
Comment (fazledyn-or) [reply spectral_cube/tests/test_spectral_cube.py:554]: Yes.

At first I was raising the same TypeError as others but running the test suite resulted in failure saying that the test "DID NOT RAISE TypeError".

I think the __floordiv__ operation is taking place here. I'm still unsure about the reason.

[the remark this answers is not in the document]
Comment (keflavich) [reply spectral_cube/tests/test_spectral_cube.py:554]: 🤯 

Alright I'll look into this more if I can - this seems like you might have uncovered a weird bug or corner case we don't understand yet.

**Comment by fazledyn-or** [comment]
I found the problem. Let's discuss it using an example-
```python
import numpy as np
from astropy.units.quantity import Quantity
from astropy.wcs import WCS
from astropy import units as u
from spectral_cube.spectral_cube import BaseSpectralCube

data = np.arange(4)[:, None, None, None] * np.ones((5, 20, 30))
wcs = WCS(naxis=3)
wcs.wcs.ctype = ['RA---TAN', 'DEC--TAN', 'FREQ']

cube = BaseSpectralCube(data=data[0], wcs=wcs)
quant = Quantity(value=1.0)
out = cube // quant
print(type(out))
```

After running this program, we get output, meaning that the `__floordiv__` operation for `BaseSpectralCube` is sucessful. The same happens for `SpectralCube` anyway.

After using `pdb` like below-
```python
# ...
quant = Quantity(value=1.0)
import pdb
pdb.set_trace()

out = cube // quant
print(type(out))
```

We can see that after getting returned `NotImplemented` from `__floordiv__` method, the operation is being handled by `__array_ufunc__` method of `astropy.units.quantity.py` at line 621.
```bash
> /home/ataf/BFAS/Source/spectral-cube/floordiv_test.py(22)<module>()
-> out = cube // quant
(Pdb) s
--Call--
> /home/ataf/BFAS/Source/spectral-cube/spectral_cube/spectral_cube.py(2298)__floordiv__()
-> def __floordiv__(self, value):
(Pdb) s
> /home/ataf/BFAS/Source/spectral-cube/spectral_cube/spectral_cube.py(2299)__floordiv__()
-> print("return NotImplemented")
(Pdb) s
return NotImplemented
> /home/ataf/BFAS/Source/spectral-cube/spectral_cube/spectral_cube.py(2300)__floordiv__()
-> return NotImplemented
(Pdb) s
--Return--
> /home/ataf/BFAS/Source/spectral-cube/spectral_cube/spectral_cube.py(2300)__floordiv__()->NotImplemented
-> return NotImplemented
(Pdb) s
--Call--
> /home/ataf/BFAS/Source/spectral-cube/env/lib/python3.8/site-packages/astropy/units/quantity.py(621)__array_ufunc__()
-> def __array_ufunc__(self, function, method, *inputs, **kwargs):
(Pdb) s
> /home/ataf/BFAS/Source/spectral-cube/env/lib/python3.8/site-packages/astropy/units/quantity.py(640)__array_ufunc__()
```

I think it has something to do with [Numpy's Behavior in combination with Python’s binary operations](https://numpy.org/neps/nep-0013-ufunc-overrides.html#behavior-in-combination-with-python-s-binary-operations). After putting print statement inside `__array_ufunc__`  method of `astropy.units.quantity.py`, we can see that `function = <ufunc 'floor_divide'>` and `method = __call__` are being passed. Since the `Quantity` class (more like numpy) is taking care of the `__floordiv__` operation here, even putting `NotImplemented` in `SpectralCube` isn't stopping the operation.

### So, why was it working before?
Before these changes, the `__floordiv__` method of `BaseSpectralCube` used to "raise" `NotImplementedError` exception which stopped the operation there. By putting `NotImplemented` instead, we allowed the Python interpreter to try out more options.

File: spectral_cube/tests/test_spectral_cube.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fazledyn-or) [reply spectral_cube/tests/test_spectral_cube.py:554]: Please have a look at https://github.com/radio-astro-tools/spectral-cube/pull/896#issuecomment-1846927831

**Comment by keflavich** [comment]
Thank you for the deep dive @fazledyn-or!  Sorry this turned out not to be the right solution, but I learned something and this is a useful record of specifically why we need to use this nonstandard approach.
